src/System.Diagnostics.DiagnosticSource/src/HttpCorrelationProtocol.md

```diff
@@ -0,0 +1,83 @@
+
+# HTTP Protocol proposal
+| Header name           |  Format    | Description |
+| ----------------------| ---------- | ---------- |
+| Request-Id            | Required. String | Unique identifier for every HTTP request involved in operation processing |
+| Correlation-Context   | Optional. Comma separated list of key-value pairs: Id=id, key1=value1, key2=value2 | Operation context which is propagated across all services involved in operation processing |
+
+## Request-Id
+`Request-Id` uniquely identifies every HTTP request involved in operation processing. 
+
+Request-Id is generated on the caller side and passed to callee. 
+
+Implementation of this protocol should expect to receive `Request-Id` in header of incoming request. 
+Absence of Request-Id indicates that it is either first instrumented service in the system or this request was not traced by upstream service and therefore does not have any context associated with it.
+To start tracing the request, implementation MUST generate new `Request-Id` (see [Root Request Id Generation](#root-request-id-generation)) for the incoming request.
+
+When Request-Id is provided by upstream service, there is no guarantee that it is unique within the entire system. 
+Implementation SHOULD make it unique by adding small suffix to incoming Request-Id to represent internal activity and use it for outgoing requests, see more details in [Hierarchical Request-Id document](HierarchicalRequestId.md).
+
+`Request-Id` is required field, which means that every instrumented request MUST have it. If implementation does not find `Request-Id` in the incoming request headers, it should consider it as non-traced and MAY not look for `Correlation-Context`.
+
+It is essential that 'incoming' and 'outgoing' Request-Ids are included in the telemetry events, so implementation of this protocol MUST provide read access to Request-Id for logging systems.
+
+### Request-Id Format
+`Request-Id` is a string up to 1024 bytes length. It contains only [Base64](https://en.wikipedia.org/wiki/Base64) and "-" (hyphen), "|" (vertical bar), "." (dot), and "_" (underscore) characters.
+
+Vertical bar, dot and underscore are reserved characters that used to mark and delimit hierarchical Request-Id, and must not be present in the nodes. Hyphen may be used in the nodes.
+
+Implementations SHOULD support hierarchical structure for the Request-Id, described in [Hierarchical Request-Id document](HierarchicalRequestId.md).
+See [Flat Request-Id](FlatRequestId.md) for non-hierarchical Request-Id requirements.
+
+## Correlation-Context
```

Comment:
This is mostly meant to control logging: service may pass sampling rate/flags there that could be used by downstream services. It's also valid to propagate other tracing-related information: feature flags, some Ids that could be useful to group traces by.

We discourage anyone from using it for non-tracing purposes. Headers have significant size limitations, parsing, propagating and just storing Correlation-Context in memory involved performance overhead. 
Service in general does not know which web-server is used somewhere downstream and exceeding header size limit may create a problem that is hard to find.

Also, depending on the implementation, service may have little control or no control at all over where Correlation-Context is propagated to: it could be external systems, cloud storage, etc. It also creates security concerns.

And finally, there are better ways and cleaner designs to pass data between the services :)